Master

danieldauk opened the pull request:
Submitting task. 
Live version: tesonet-test.surge.sh

danieldauk committed:
app setup

danieldauk committed:
authorized server list fetching, server component, servers container, servers sorting

danieldauk committed:
implemented logout (Router problems?)

danieldauk committed:
fixed router problem

danieldauk committed:
non-responsive login style

danieldauk committed:
servers styling start

danieldauk committed:
servers styling end. Deploy test.

danieldauk committed:
IE11 support

danieldauk committed:
reducer tests

danieldauk committed:
Servers container tests

danieldauk committed:
App tests

danieldauk committed:
spinner during authorization

danieldauk committed:
servers spinner

danieldauk committed:
servers error message

danieldauk committed:
additional tests. Finished.

Mindaugas-Jacionis commented:
Hey @danieldauk - I am impressed with your solution. It is great that you have some test coverage! 

I have few requests:
- Could you add `eslint` to your project? Reason: it is good to follow standarts, but it is even better to have tooling that keeps you on track. I'd recommend to use `airbnb` config(you can feel free to do some modifications)
- Add `prettier` to your project. This will help out with code formatting and will enforce `eslint`.
- Could you move store creation to `src/store/index.js` and import it from there? This creates separation of concerns and when app gets bigger it is easier to maintain.
- Could you move the `Provider` && `BrowserRouter` to `App`?  Your entry point(`index.js`) should be as lean as possible.

Few questions:

Mindaugas-Jacionis commented:
-

Mindaugas-Jacionis commented:
Hey @danieldauk - I am impressed with your solution. It is great that you have some test coverage! 

I have few requests:
- Could you add `eslint` to your project? Reason: it is good to follow standarts, but it is even better to have tooling that keeps you on track. I'd recommend to use `airbnb` config(you can feel free to do some modifications)
- Add `prettier` to your project. This will help out with code formatting and will enforce `eslint`.
- Could you move store creation to `src/store/index.js` and import it from there? This creates separation of concerns and when app gets bigger it is easier to maintain.
- Could you move the `Provider` && `BrowserRouter` to `App`?  Your entry point(`index.js`) should be as lean as possible.

Few questions:
- How you would restructure your redux state implementation when app would be much larger with possibly a lot of different api calls? Single file solution like now wouldn't be viable, because it would easily become a mess.
- Why you chose to use `.png` for login icons?
- Would it make sense to structure your `assets` directory?
- Is there any specific reason you decided to use pure `css` instead using preprocessors like `Sass` or `Less`?
- Why no tests on components?
- Are you using service worker for something or this is leftover from `create-react-app`?

Comments:
- If you have dedicated directory for component/container - there is no need to name the file like component/container, it is better to use `index.js` and `index.css`. So instead of `src/containers/Login/Login.js`, it better be `src/containers/Login/index.js`. This is not a mistake and totally viable solution, but it helps with structure, imports(instead of `import Login from './containers/Login/Login';` it is enough to write `import Login from './containers/Login';`) and search in editor is better then.
- You could have `index.js` in your `components` and `contianers` directories that re-export your components/containers. Then imports can look something like this: `import { .Login, Servers } from './containers';` instead of multiple lines with specific directories.

There are further comments in the code, to help you improve.

Recommendations:
- Use [prop-types](https://www.npmjs.com/package/prop-types) to validate that properties that are passed to your components and containers are of valid type. This will help in debugging and maintaining your codebase a lot. You could of course use `flow` or `typescript` to go even further on type checking, but IMHO those are too intrusive - at least for people that are just starting out, so I wouldn't worry about those for now.
- Use `yarn` for packages, it is faster than `npm` and uses same registry, so you don't need to change any files, just install another tool and use it.

Overall you did a great job, as for junior developer - your knowledge of react and its tooling is really good. You should be proud of your job. 👍 

Cheers! :)

Mindaugas-Jacionis commented on src/App.css:
```diff
@@ -0,0 +1,20 @@
+  width: 100%;
+}
+
+::-webkit-scrollbar {
```
I'd suggest to avoid editing scrollbar appearance. It is very cross browser and cross platform inconsistent.

Mindaugas-Jacionis commented on src/App.js:
```diff
@@ -0,0 +1,53 @@
+  render() {
+    //guard routes depending on login state
+
+    let routes = (
```
It is bad practice to create components inside your `render` method. Better create separate methods that return the view and call them.

Further, here you have couple of options, one is that your render could look something like this:
```javascript
render() {
    const hasToken = !!localStorage.getItem('token');

    return (
      <div className="App">
        <Switch>
          <Route path="/" exact component={hasToken ? Servers : Login} />
          <Redirect to="/" />
        </Switch>
      </div>
    );
  }
}
```
Another is to have:
```javascript
render() {
    const hasToken = !!localStorage.getItem('token');

    return (
      <div className="App">
        <Switch>
          <Route path="/login" exact component={Login} />
          <Route path="/servers" exact component={Servers} />
          <Redirect to={hasToken ? '/servers' : '/login'} />
        </Switch>
      </div>
    );
  }
}
```

Second option is a bit flawed, because user can access pages via direct link, so you would need to validate stuff in `componentDidMount`. So second option can be improved using a technique called `ProtectedRoute`. You can read more about it [in this blog post](https://medium.com/@s4y.solutions/react-route-4-protected-route-even-simpler-9b89dc129cde), [and this one](https://tylermcginnis.com/react-router-protected-routes-authentication/).

Mindaugas-Jacionis commented on src/components/Header/Header.js:
```diff
@@ -0,0 +1,20 @@
+import "./Header.css";
+import LogoutIcon from "../LogoutIcon/LogoutIcon";
+
+const header = props => {
```
It is a component, you should be using upper case name.

Mindaugas-Jacionis commented on src/components/LogoutIcon/LogoutIcon.js:
```diff
@@ -0,0 +1,20 @@
+import React from "react";
+
+const logoutIcon = () => {
```
If you would have more icons, how would you go about that? Would you have more components like this one? :)

Mindaugas-Jacionis commented on src/components/Server/Server.css:
```diff
@@ -0,0 +1,9 @@
+.server-container{
```
Recommendation: read-up on [BEM](http://getbem.com/introduction/), it is a great way to approach css modularity(since React by nature is modular, it only makes sense to have modular styling)

Mindaugas-Jacionis commented on src/components/Server/Server.js:
```diff
@@ -0,0 +1,20 @@
+import "./Server.css";
+
+const server = (props) =>{
+
```
This whitespace shouldn't be here. `eslint` and `prettier` will help you to avoid this.

Mindaugas-Jacionis commented on src/components/Server/Server.js:
```diff
@@ -0,0 +1,20 @@
+    return (
+        <div className="server-container">
+            <div>
+                {props.name}
```
It is never a good idea to use plain text in block elements. You should wrap your text in `p`, `h` or other text tag. It works, but semantically it is wrong.

Mindaugas-Jacionis commented on src/components/Spinner/Spinner.js:
```diff
@@ -0,0 +1,24 @@
+    marginTop: props.marginTop
+  }
+
+  return (
```
Just an improvement idea: you could have couple sizes(i.e. `small`, `medium`, `large`) and modes(`default` and `fullscreen`) and accordingly apply styling and container to (not)cover whole screen. :)

Mindaugas-Jacionis commented on src/containers/Servers/Servers.js:
```diff
@@ -0,0 +1,89 @@
+
+  componentWillReceiveProps(nextProps) {
+    if (nextProps.token === "") {
+      this.props.history.push("/");
```
I'd suggest using `replace`, so user couldn't use back button to return to the screen that requires auth token.

Mindaugas-Jacionis commented on src/containers/Servers/Servers.js:
```diff
@@ -0,0 +1,89 @@
+    }
+
+    if (this.props.servers) {
+      servers = this.props.servers.map(server => {
```
move this to `renderServers` method

Mindaugas-Jacionis commented on src/containers/Servers/Servers.js:
```diff
@@ -0,0 +1,89 @@
+    let servers = null;
+    
+    if (this.props.loading) {
+      servers = <Spinner marginTop="30px" color="#9fd533" />;
```
move this to `renderLoading` method
plus it is a bad naming, variable is named servers, but you are putting loader in it
since you are reusing same variable it would make sense to call it `view` or `content`

Mindaugas-Jacionis commented on src/containers/Servers/Servers.js:
```diff
@@ -0,0 +1,89 @@
+    }
+
+    if (this.props.error) {
+      servers = (
```
move this to `renderError` method

Mindaugas-Jacionis commented on src/containers/Servers/Servers.js:
```diff
@@ -0,0 +1,89 @@
+      <div className="servers-container">
+        <Header logout={this.props.logout} />
+        <div className="servers-list-header">
+          <p>SERVERS</p>
```
You can use `css` to enforce uppercase here. [Resource](https://css-tricks.com/almanac/properties/t/text-transform/)

Mindaugas-Jacionis commented on src/containers/Servers/Servers.js:
```diff
@@ -0,0 +1,89 @@
+          <p>SERVERS</p>
+          <p>DISTANCE</p>
+        </div>
+        <div className="servers-list">{servers}</div>
```
Better:
```javascript
{this.props.loading && this.renderLoading}
{this.props.error ? this.renderError : this.renderServer}}
```

Mindaugas-Jacionis commented on src/containers/Login/Login.js:
```diff
@@ -0,0 +1,79 @@
+    this.setState({ username: event.target.value });
+  };
+
+  passwordChangeHandler = event => {
```
you can combine `usernameChangeHandler` and `passwordChangeHandler` to one function:
```javascript
// you pass in field name
onInput = (event, field) => this.setStat({ [field]: event.target.value });
```

Mindaugas-Jacionis commented on src/containers/Login/Login.js:
```diff
@@ -0,0 +1,79 @@
+              type="password"
+              required
+            />
+            <button>{this.props.loading? <Spinner/>: "Log in"}</button>
```
button type should be `submit` - semantics of the form require this

Mindaugas-Jacionis commented on src/containers/Servers/Servers.js:
```diff
@@ -0,0 +1,89 @@
+    }
+  }
+
+  render() {
```
when you use more than one value from `props` or `state` - it makes sense to destructure them at the beginning of the method.
So, you can do this here:
`const { loading, error, servers, logout } = this.props;`

And then instead of `this.props.loading` you can simply write `logout` - code becomes less verbose and easier to read. :)

danieldauk committed:
amendments

Added eslint and prettier. Moved store creation to src/store/index.js.
Moved Provider and BrowserRouter to App. Added props validation and
default props. Structured assets directory. Renamed
components/containers files to index.js. Re-exported
containers/components from index.js file.

danieldauk committed:
svg icons instead of png

danieldauk committed:
test changes. Additional components tests.

danieldauk committed:
eslint jsx-filename-extension rule

danieldauk commented:
Hi @Mindaugas-Jacionis,

First of all, I would like to thank you for your time and detailed feedback. I really appreciate it. 

**Regarding requests:**
- `Eslint` added with airbnb config. Made some config changes to allow using `JSX `in `.js` files and rule regarding use of stateless function ( I need stateful components to use prop-types).
- `Prettier` added to project.
- Store creation moved to `src/store/index.js`.
- `Provider` and `Browser Router` moved to App. 

**Regarding questions:**

- I would restructure redux state as follows: I would split reducers according to it's "function in code". For example, one reducer would be responsible for authentication another for servers, etc. Then we can combine reducers using `combineReducers` from "redux. Same with actions. Although combining actions would be a little bit different: we can have one `index.js` file where we re-export actions
- good questions :) It is better to use `SVG` files because we can style them with `CSS` and resolution is not limited. Project modified - now I'm using `SVG`.
- It would make a lot of sense to structure assets directory. Done.
- I've decided to use pure `CSS` because the benefits of using preprocessors in such a small project would be minimal. 
- Some components are just returning `JSX` with no dependency on external props, so I think it is not necessary to test them. Regarding components that depend on external props - I've corrected my mistake and written tests for them.
- To be honest, I'm not very familiar with service worker. So your assumption is correct - it is a leftover from `create-react-app`.

**Regarding comments:**

- All components/containers file names renamed to `index.js` and` index.css`.
- Made an `index.js` file and re-exported components/ containers. 
- These approaches are more convenient. Thank you :)

**Regarding recommendations:**

- added prop validation and prop defaults to project.

danieldauk committed:
amendments 2

Removed scrollbar styling.
Separate methods for different routes depending on login state.
Changed svg components to sprite.

danieldauk committed:
amendments 3

Html semantics.
Combined inputChangeHandler.
Added renderServers, renderLoading, render Error methods.

danieldauk committed:
spinner modes and spinner tests

danieldauk committed:
BEM

danieldauk committed:
spinner IE11 fix. Tests fix.

danieldauk committed:
IE11 external svg fix

danieldauk committed:
IE11 external svg fix
